cli/src/commands/rebase.rs

```diff
@@ -443,6 +462,160 @@ fn rebase_revision(
     }
 }
 
+/// Extracts `target_commits` from the graph by rebasing its descendants onto
+/// its parents. This assumes that `target_commits` can be rewritten.
+/// `target_commits` should be in reverse topological order.
+/// Returns a map from the original commit IDs to the rewritten commit IDs, and
+/// the roots of the `target_commits` set, using the rewritten commit ID if
+/// applicable.
+fn extract_commits(
+    tx: &mut WorkspaceCommandTransaction,
+    settings: &UserSettings,
+    target_commits: &[Commit],
+) -> Result<(HashMap<CommitId, CommitId>, HashSet<CommitId>), CommandError> {
+    let target_commit_ids: HashSet<_> = target_commits.iter().ids().cloned().collect();
+
+    let repo = tx.mut_repo();
+    let connected_target_commits: Vec<_> =
+        RevsetExpression::commits(target_commits.iter().ids().cloned().collect_vec())
+            .connected()
+            .evaluate_programmatic(repo)?
+            .iter()
+            .commits(repo.store())
+            .try_collect()?;
+
+    // Construct a list of parents for each commit in the connected graph of the
+    // target set which is not in the target set, which consists of ancestors which
+    // are in the target set.
+    let mut connected_target_parents: HashMap<CommitId, IndexSet<CommitId>> = HashMap::new();
+    for commit in connected_target_commits.iter().rev() {
+        if target_commit_ids.contains(commit.id()) {
+            continue;
+        }
+        let mut new_parents = IndexSet::new();
+        for old_parent in commit.parent_ids() {
+            if let Some(parents) = connected_target_parents.get(old_parent) {
+                new_parents.extend(parents.iter().cloned());
+            } else if target_commit_ids.contains(old_parent) {
+                new_parents.insert(old_parent.clone());
+            }
+        }
+        connected_target_parents.insert(commit.id().clone(), new_parents);
+    }
+
+    // Commits in the target set should only have other commits in the set as
+    // parents, except the roots of the set, which persist their original
+    // parents.
+    let mut new_target_parents: HashMap<CommitId, Vec<CommitId>> = HashMap::new();
+    for commit in target_commits.iter().rev() {
+        // The roots of the set will not have any parents found in `new_target_parents`,
+        // and will be stored in `new_target_parents` as an empty vector.
+        let mut new_parents = vec![];
+        for old_parent in commit.parent_ids() {
+            if new_target_parents.contains_key(old_parent) {
+                new_parents.push(old_parent.clone());
+            } else if let Some(parents) = connected_target_parents.get(old_parent) {
+                new_parents.extend(parents.iter().cloned());
+            }
+        }
+        new_target_parents.insert(commit.id().clone(), new_parents);
+    }
+
```

Comment:
Perhaps, the first two loops can be combined as
```rust
    let mut new_target_parents: HashMap<CommitId, Vec<CommitId>> = HashMap::new();
    for commit in connected_target_commits.iter().rev() {
        // The roots of the set will not have any parents found in `new_target_parents`,
        // and will be stored in `new_target_parents` as an empty vector.
        let mut new_parents = vec![];
        for old_parent in commit.parent_ids() {
            if target_commit_ids.contains(old_parent) {
                new_parents.push(old_parent.clone());
            } else if let Some(parents) = new_target_parents.get(old_parent) {
                new_parents.extend(parents.iter().cloned());
            }
        }
        new_target_parents.insert(commit.id().clone(), new_parents);
    }
    new_target_parents.retain(|id, _| target_commit_ids.contains(id)); // or test membership separately
```

Reply:
Good suggestion!